### ISSUE TITLE
Rule Change: no-constant-condition should ignore `while (true)`

### ISSUE THREAD

**Issue by branko-d** [eslint/eslint#17807]
### What rule do you want to change?

no-constant-condition

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
let condition = false;

while (true) {
    // ...
    if (!condition)
        break;
    // ...
}

while (typeof condition) {
    // ...
}
```

### What does the rule currently do for this code?

Treats both loops the same, either fails them both or passes them both (depending on `checkLoops`).

### What will the rule do after it's changed?

Treat the loops differently:

- pass the first loop
- and fail the second - the programmer probably didn't want `typeof` there.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

Since `while (true)` is a perfectly legitimate way to write a loop, it should work out-of-box. Alternatively, we should have an option to ignore just `while (true)` without also ignoring loops that might be genuinely problematic.

**Comment by mdjermanovic** [comment]
> Alternatively, we should have an option to ignore just `while (true)` without also ignoring loops that might be genuinely problematic.

I support this proposal. I agree that ignoring all loops by setting `checkLoops:false`  just to allow `while(true)`, which is a common idiom and does not in itself imply that the code is invalid, is probably not what most users want.

Waiting for opinions from other team members.

**Comment by fasttime** [comment]
The alternative to `while (true)` that doesn't use a constant expression is `for (;;)`, but it's not obvious, and the fact that this rule does not provide suggestions doesn't help. So I think it's good to ignore `while (true)` by default. If we could also have an option to decide whether to ignore it or not, it would be better.

> In my opinion, the `checkLoops: false` option is too loose a check for me. If we want the `no-constant-condition` rule to ignore `while (true)` by default, I want the `no-constant-condition` rule to check if the `while (true)` statement has a `break` or `return`.

I'm not sure about this. Some loops can be terminated by an exception, or by `continue` with a label, or `process.exit()` on Node.js. [Infinite generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator?retiredLocale=vi#an_infinite_iterator) with endless loops are nothing exotic. There are many cases.

I can totally see why it would be helpful to ensure that a loop has a more-or-less canonical exit path, even if its main condition is constant. But determining whether that is the case or not is maybe too much of a stretch for a simple rule like this. It feels like a different problem, to me at least.

**Comment by mdjermanovic** [comment]
I also think that checking whether a loop has an exit path doesn't seem like it should be the responsibility of this rule.

**Comment by ota-meshi** [comment]
I see. You are right that checking things like `break` statements should not be implemented with `no-constant-condition` rule.
Do you think it will be accepted if I open another rule proposal issue?

**Comment by Rec0iL99** [comment]
Not stale. Looks like we have the consensus here to move this forward to ignore `while (true)` by default for `no-constant-condition`

**Comment by fasttime** [comment]
I was thinking that we could extend the value range of the `checkLoops` option (currently a boolean) to describe one of three options, for example:
- `"always"` or `true`: report constant expressions in loops. This is the current default.
- `"unlessTrue"`: report constant expressions in loops but allow the literal `true`. This would be the new default.
- `"never"` or `false`: do not report constant expressions in loops.

**Comment by Tanujkanti4441** [comment]
> I was thinking that we could extend the value range of the `checkLoops` option (currently a boolean) to describe one of three options, for example:
> 
>     * `"always"` or `true`: report constant expressions in loops. This is the current default.
> 
>     * `"unlessTrue"`: report constant expressions in loops but allow the literal `true`. This would be the new default.
> 
>     * `"never"` or `false`: do not report constant expressions in loops.

that is really a good idea but this issue only focuses on `while` loop how about if we will have the following value of `checkloop`:
`all`: report all loops with constant conditions,
`allExceptWhile`: report all loops except `while` or `do while` loops with `true` literal.
`none`: do not report loops with constant conditions.

**Comment by Tanujkanti4441** [comment]
ping @eslint/eslint-team

**Comment by nzakas** [comment]
I agree with @Tanujkanti4441 -- we are looking to focus specifically on `while(true)` and not other loops (there was no mention of `for(;true;)` or `do { } while(true)` with `true` condition, so a more focused option name is important. So maybe along the same lines:

* `all` - report all loops
* `allExceptWhileTrue` - report all loops except `while(true)`
* `none` - don't report on loops

**Comment by fasttime** [comment]
> I agree with @Tanujkanti4441 -- we are looking to focus specifically on `while(true)` and not other loops (there was no mention of `for(;true;)` or `do { } while(true)` with `true` condition, so a more focused option name is important. So maybe along the same lines:
> 
>     * `all` - report all loops
>     * `allExceptWhileTrue` - report all loops except `while(true)`
>     * `none` - don't report on loops

Okay, let's do it like this, marking as accepted.

So we have agreed to extend the `checkLoops` option with three new values: `all` (like `true`), `none` (like `false`), and `allExceptWhileTrue` which will behave like `all` but will not flag `while (true)`. `allExceptWhileTrue` will be the new default.

@branko-d would you be availble to work on a pull request for this change?

**Comment by Tanujkanti4441** [comment]
I'll take this.
